sklearn/utils/estimator_checks.py

```diff
@@ -454,8 +455,37 @@ def check_dict_unchanged(name, Estimator):
                               'Estimator changes __dict__ during %s' % method)
 
 
+def check_fit_adds_private_attributes_only(name, Estimator):
+    rnd = np.random.RandomState(0)
+    X = 3 * rnd.uniform(size=(20, 3))
+    y = X[:, 0].astype(np.int)
+    y = multioutput_estimator_convert_y_2d(name, y)
+    estimator = Estimator()
+    set_testing_parameters(estimator)
+
+    if hasattr(estimator, "n_components"):
+        estimator.n_components = 1
+    if hasattr(estimator, "n_clusters"):
+        estimator.n_clusters = 1
+
+    set_random_state(estimator, 1)
+    dict_before_fit = estimator.__dict__.copy()
+    estimator.fit(X, y)
+
+    dict_after_fit = estimator.__dict__
+    # leave only attributes which have been set by fit
```

Comment:
haha I understand! Just trying to figure out how to make this run and for me your sentence saying: "attributes set in `__init__` were not modified by `fit`" and  @amueller comment some time ago saying: "We certainly change the `__dict__` during `fit`" conflicting in my mind, so I can't really continue here without resolving this conflict.

So the real question is: should we or shouldn't we modify `__dict__` during `fit` for any given estimator?

Do I understand correctly, that we can add new attributes to `__dict__` during `fit`, but we can't modify the old one (set by `__init__`)?